File: VB/ReorderHelper.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraPivotGrid
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms

Namespace Pivot_Reorder
	Friend Class ReorderHelper
		Private targetPivot_Renamed As PivotGridControl
		Private ReadOnly Property TargetPivot() As PivotGridControl
			Get
				Return targetPivot_Renamed
			End Get
		End Property
		Private privateAllowChangeRowGroup As Boolean
		Public Property AllowChangeRowGroup() As Boolean
			Get
				Return privateAllowChangeRowGroup
			End Get
			Set(ByVal value As Boolean)
				privateAllowChangeRowGroup = value
			End Set
		End Property
		Private privateAllowChangeColumnGroup As Boolean
		Public Property AllowChangeColumnGroup() As Boolean
			Get
				Return privateAllowChangeColumnGroup
			End Get
			Set(ByVal value As Boolean)
				privateAllowChangeColumnGroup = value
			End Set
		End Property
		Private _allowDragRow As Boolean
		Public Property AllowDragRow() As Boolean
			Get
				Return _allowDragRow
			End Get
			Set(ByVal value As Boolean)
				If _allowDragRow = value Then
					Return
				End If
				_allowDragRow = value
				If _allowDragRow Then
					InitRowsOrder()
				End If
			End Set
		End Property
		Private _allowDragColumn As Boolean
		Public Property AllowDragColumn() As Boolean
			Get
				Return _allowDragColumn
			End Get
			Set(ByVal value As Boolean)
				If _allowDragColumn = value Then
					Return
				End If
				_allowDragColumn = value
				If _allowDragColumn Then
					InitColumnsOrder()
				End If
			End Set
		End Property


		Public Sub New(ByVal pivot As PivotGridControl)
			AllowChangeRowGroup = False
			AllowChangeColumnGroup = False
			AllowDragColumn = False
			Attach(pivot)
		End Sub
		Public Sub Attach(ByVal pivot As PivotGridControl)
			If targetPivot_Renamed Is pivot Then
				Return
			ElseIf targetPivot_Renamed IsNot Nothing Then
				Detach(targetPivot_Renamed)
			End If

			targetPivot_Renamed = pivot
			AddHandler targetPivot_Renamed.CustomFieldSort, AddressOf targetPivot_CustomFieldSort
			AddHandler targetPivot_Renamed.MouseMove, AddressOf targetPivot_MouseMove
			AddHandler targetPivot_Renamed.MouseUp, AddressOf targetPivot_MouseUp
			AddHandler targetPivot_Renamed.FieldAreaChanged, AddressOf targetPivot_FieldAreaChanged
			AddHandler targetPivot_Renamed.CustomDrawFieldValue, AddressOf targetPivot_CustomDrawFieldValue
		End Sub

		Private Sub targetPivot_CustomDrawFieldValue(ByVal sender As Object, ByVal e As PivotCustomDrawFieldValueEventArgs)
			If onDrag Then
				'if (e.ValueType != PivotGridValueType.Value) return;
				'if(e.Field ==)
                If e.Field Is dragInfo.ValueInfo.Field AndAlso e.Bounds.Contains(dragInfo.HitPoint) Then
                    e.Painter.DrawObject(e.Info)
                    e.Handled = True
                    Dim rec As Rectangle = e.Info.CaptionRect
                    e.Graphics.DrawRectangle(Pens.Blue, rec)
                    Return
                End If
                If TargetInfo IsNot Nothing AndAlso e.Field Is TargetInfo.ValueInfo.Field AndAlso e.Bounds.Contains(TargetInfo.HitPoint) Then
                    e.Painter.DrawObject(e.Info)
                    e.Handled = True
                    Dim rec As Rectangle = e.Info.CaptionRect
                    e.Graphics.DrawRectangle(Pens.Green, rec)
                    Return
                End If
			End If
			'throw new NotImplementedException();
		End Sub

		Private Sub targetPivot_FieldAreaChanged(ByVal sender As Object, ByVal e As PivotFieldEventArgs)
			If (e.Field.Area = PivotArea.ColumnArea AndAlso AllowDragColumn) OrElse (e.Field.Area = PivotArea.RowArea AndAlso AllowDragRow) Then
				InitSingleFieldOrder(e.Field)
			End If
		End Sub

		Private Sub InitRowsOrder()
			For Each field As PivotGridField In targetPivot_Renamed.GetFieldsByArea(PivotArea.RowArea)
				InitSingleFieldOrder(field)
			Next field
		End Sub
		Private Sub InitSingleFieldOrder(ByVal field As PivotGridField)
			Dim allowedValues() As Object = field.GetUniqueValues()
			'IEnumerable<object> sortedGroupValues = allowedValues.OrderBy(value => value.ToString());
			Dim orderList As New List(Of Object)()
			'foreach (object groupValue in sortedGroupValues)
			For Each groupValue As Object In allowedValues
				orderList.Add(groupValue)
			Next groupValue
			field.Tag = orderList
			field.SortMode = PivotSortMode.Custom
		End Sub

		Private Sub RemoveSingleFieldOrder(ByVal field As PivotGridField)
			field.SortMode = PivotSortMode.Default
		End Sub

		Private Sub InitColumnsOrder()
			For Each field As PivotGridField In targetPivot_Renamed.GetFieldsByArea(PivotArea.ColumnArea)
				InitSingleFieldOrder(field)
			Next field
		End Sub
		Public Sub Detach(ByVal pivot As PivotGridControl)
			For Each field As PivotGridField In targetPivot_Renamed.Fields
					field.Tag = Nothing
					field.SortMode = PivotSortMode.Default
			Next field

		End Sub
		Private Sub targetPivot_CustomFieldSort(ByVal sender As Object, ByVal e As DevExpress.XtraPivotGrid.PivotGridCustomFieldSortEventArgs)
			If AllowDragRow AndAlso e.Field.Area = PivotArea.RowArea Then
				Dim orderlist As List(Of Object) = TryCast(e.Field.Tag, List(Of Object))
				'Comparer comp = new Comparer;
				If orderlist Is Nothing OrElse orderlist.IndexOf(e.Value1) = -1 OrElse orderlist.IndexOf(e.Value2) = -1 Then
					'e.Result = Comparer.DefaultInvariant.Compare(e.Value1, e.Value2);
					'e.Handled = true;
					Return
				End If
				e.Result = Comparer.DefaultInvariant.Compare(orderlist.IndexOf(e.Value1), orderlist.IndexOf(e.Value2))
				e.Handled = True
			End If
			If AllowDragColumn AndAlso e.Field.Area = PivotArea.ColumnArea Then
				Dim orderlist As List(Of Object) = TryCast(e.Field.Tag, List(Of Object))
				If orderlist Is Nothing OrElse orderlist.IndexOf(e.Value1) = -1 OrElse orderlist.IndexOf(e.Value2) = -1 Then
					Return
				End If
				e.Result = Comparer.DefaultInvariant.Compare(orderlist.IndexOf(e.Value1), orderlist.IndexOf(e.Value2))
				e.Handled = True
			End If
		End Sub
		Private dragInfo As PivotGridHitInfo
		Private _targetInfo As PivotGridHitInfo
		Private Property TargetInfo() As PivotGridHitInfo
			Get
				Return _targetInfo
			End Get
			Set(ByVal value As PivotGridHitInfo)
				_targetInfo = value
			End Set
		End Property
		Private onDrag As Boolean = False
		Private Sub targetPivot_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
			If (Not onDrag) Then
				If e.Button = System.Windows.Forms.MouseButtons.Left Then
					Dim hi As PivotGridHitInfo = targetPivot_Renamed.CalcHitInfo(e.Location)
					If hi.HitTest = PivotGridHitTest.Value Then
						If (AllowDragRow AndAlso hi.ValueInfo.Field.Area = PivotArea.RowArea) OrElse (AllowDragColumn AndAlso hi.ValueInfo.Field.Area = PivotArea.ColumnArea) Then
							dragInfo = hi
							onDrag = True
							targetPivot_Renamed.FindForm().Cursor = Cursors.Hand
						End If
					End If
				End If
			Else
				Dim allowedEffect As DragDropEffects = GetDragEffect(e.Location)
				If allowedEffect = DragDropEffects.None Then
					targetPivot_Renamed.FindForm().Cursor = Cursors.No
				Else
					targetPivot_Renamed.FindForm().Cursor = Cursors.Hand
				End If


			End If

		End Sub

		Private Sub targetPivot_MouseUp(ByVal sender As Object, ByVal e As MouseEventArgs)
			If onDrag Then
				If GetDragEffect(e.Location) = DragDropEffects.Move Then
					Dim hi As PivotGridHitInfo = targetPivot_Renamed.CalcHitInfo(e.Location)
                    If hi.ValueInfo.Field Is dragInfo.ValueInfo.Field Then
                        TryCast(dragInfo.ValueInfo.Field.Tag, IList).Remove(dragInfo.ValueInfo.Value)
                        TryCast(dragInfo.ValueInfo.Field.Tag, IList).Insert((TryCast(dragInfo.ValueInfo.Field.Tag, IList)).IndexOf(hi.ValueInfo.Value), dragInfo.ValueInfo.Value)
                        'updateMemos();
                    End If
					If (dragInfo.ValueInfo.Field.Area = PivotArea.RowArea AndAlso AllowChangeRowGroup) OrElse (dragInfo.ValueInfo.Field.Area = PivotArea.ColumnArea AndAlso AllowChangeColumnGroup) Then
						Dim ds As PivotDrillDownDataSource = Nothing
						Dim i As Integer = 0
						Do While i < dragInfo.ValueInfo.Field.AreaIndex AndAlso i <= hi.ValueInfo.Field.AreaIndex
							Dim targetFieldValue As Object = hi.ValueInfo.GetFieldValue(targetPivot_Renamed.GetFieldByArea(dragInfo.ValueInfo.Field.Area, i), hi.ValueInfo.MinIndex)
							Dim dragFieldValue As Object = dragInfo.ValueInfo.GetFieldValue(targetPivot_Renamed.GetFieldByArea(dragInfo.ValueInfo.Field.Area, i), dragInfo.ValueInfo.MinIndex)
							If targetFieldValue IsNot dragFieldValue Then
								If ds Is Nothing Then
									ds = dragInfo.ValueInfo.CreateDrillDownDataSource()
								End If
								For Each row As PivotDrillDownDataRow In ds
									ds.SetValue(row.Index, targetPivot_Renamed.GetFieldByArea(dragInfo.ValueInfo.Field.Area, i), targetFieldValue)
								Next row
							End If
							i += 1
						Loop
					End If
					targetPivot_Renamed.RefreshData()
				End If

				onDrag = False
				targetPivot_Renamed.FindForm().Cursor = Cursors.Default
			End If
		End Sub

		Private Function GetDragEffect(ByVal point As Point) As DragDropEffects
			Dim hi As PivotGridHitInfo = targetPivot_Renamed.CalcHitInfo(point)
			Dim dragArea As PivotArea = dragInfo.ValueInfo.Field.Area
			If hi.HitTest <> PivotGridHitTest.Value OrElse hi.ValueInfo.Field Is Nothing OrElse hi.ValueInfo.Field.Area <> dragArea Then
				TargetInfo = Nothing
				Return DragDropEffects.None
			End If
			If (dragArea = PivotArea.ColumnArea AndAlso (Not AllowDragColumn)) OrElse (dragArea = PivotArea.RowArea AndAlso (Not AllowDragRow)) Then
				TargetInfo = Nothing
				Return DragDropEffects.None
			End If
			'for the same group
			If IsSameGroup(dragInfo.ValueInfo, hi.ValueInfo) Then
                If dragInfo.ValueInfo.Field IsNot hi.ValueInfo.Field OrElse dragInfo.ValueInfo.Value.Equals(hi.ValueInfo.Value) Then
                    TargetInfo = Nothing
                    Return DragDropEffects.None
                Else
                    TargetInfo = hi
                    Return DragDropEffects.Move
                End If
			Else 'for different groups
				If (dragArea = PivotArea.RowArea AndAlso AllowChangeRowGroup) OrElse (dragArea = PivotArea.ColumnArea AndAlso AllowChangeColumnGroup) Then
					TargetInfo = hi
					Return DragDropEffects.Move
				Else
					TargetInfo = Nothing
					Return DragDropEffects.None
				End If
			End If
		End Function

		Private Function IsSameGroup(ByVal dragInfo As PivotFieldValueHitInfo, ByVal targetInfo As PivotFieldValueHitInfo) As Boolean
			If dragInfo.Field.Area <> targetInfo.Field.Area Then
				Return False
			End If
			Dim minAreaIndex As Integer = Math.Min(dragInfo.Field.AreaIndex-1, targetInfo.Field.AreaIndex)
			For i As Integer = 0 To minAreaIndex
				Dim field1GroupValue As Object = dragInfo.GetFieldValue(targetPivot_Renamed.GetFieldByArea(dragInfo.Field.Area, i), dragInfo.MinIndex)
				Dim field2GroupValue As Object = targetInfo.GetFieldValue(targetPivot_Renamed.GetFieldByArea(targetInfo.Field.Area, i), targetInfo.MinIndex)
				If (Not field1GroupValue.Equals(field2GroupValue)) Then
					Return False
				End If
			Next i
			Return True
		End Function


	End Class
End Namespace

File: CS/ReorderHelper.cs
using DevExpress.XtraPivotGrid;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace Pivot_Reorder
{
    class ReorderHelper
    {
        PivotGridControl targetPivot;
        PivotGridControl TargetPivot { get { return targetPivot; } }
        public bool AllowChangeRowGroup { get; set; }
        public bool AllowChangeColumnGroup { get; set; }
        bool _allowDragRow;
        public bool AllowDragRow
        {
            get { return _allowDragRow; }
            set
            {
                if (_allowDragRow == value)
                    return;
                _allowDragRow = value;
                if (_allowDragRow)
                    InitRowsOrder();
            }
        }
        bool _allowDragColumn;
        public bool AllowDragColumn
        {
            get { return _allowDragColumn; }
            set
            {
                if (_allowDragColumn == value)
                    return;
                _allowDragColumn = value;
                if (_allowDragColumn)
                    InitColumnsOrder();
            }
        }


        public ReorderHelper(PivotGridControl pivot)
        {
            AllowChangeRowGroup = false;
            AllowChangeColumnGroup = false;
            AllowDragColumn = false;
            Attach(pivot);
        }
        public void Attach(PivotGridControl pivot)
        {
            if (targetPivot == pivot)
                return;
            else if (targetPivot != null)
                Detach(targetPivot);

            targetPivot = pivot;
            targetPivot.CustomFieldSort += targetPivot_CustomFieldSort;
            targetPivot.MouseMove += targetPivot_MouseMove;
            targetPivot.MouseUp += targetPivot_MouseUp;
            targetPivot.FieldAreaChanged += new PivotFieldEventHandler(targetPivot_FieldAreaChanged);
            targetPivot.CustomDrawFieldValue += targetPivot_CustomDrawFieldValue;
        }

        void targetPivot_CustomDrawFieldValue(object sender, PivotCustomDrawFieldValueEventArgs e)
        {
            if (onDrag)
            {
                //if (e.ValueType != PivotGridValueType.Value) return;
                //if(e.Field ==)
                if (e.Field == dragInfo.ValueInfo.Field && e.Bounds.Contains(dragInfo.HitPoint))
                {
                    e.Painter.DrawObject(e.Info);
                    e.Handled = true;
                    Rectangle rec = e.Info.CaptionRect;
                    e.Graphics.DrawRectangle(Pens.Blue, rec);
                    return;
                }
                if (TargetInfo != null && e.Field == TargetInfo.ValueInfo.Field && e.Bounds.Contains(TargetInfo.HitPoint))
                {
                    e.Painter.DrawObject(e.Info);
                    e.Handled = true;
                    Rectangle rec = e.Info.CaptionRect;
                    e.Graphics.DrawRectangle(Pens.Green, rec);
                    return;
                }
            }
            //throw new NotImplementedException();
        }

        void targetPivot_FieldAreaChanged(object sender, PivotFieldEventArgs e)
        {
            if ((e.Field.Area == PivotArea.ColumnArea && AllowDragColumn) || (e.Field.Area == PivotArea.RowArea && AllowDragRow))
                InitSingleFieldOrder(e.Field);
        }

        void InitRowsOrder()
        {
            foreach (PivotGridField field in targetPivot.GetFieldsByArea(PivotArea.RowArea))
            {
                InitSingleFieldOrder(field);
            }
        }
        void InitSingleFieldOrder(PivotGridField field)
        {
            object[] allowedValues = field.GetUniqueValues();
            //IEnumerable<object> sortedGroupValues = allowedValues.OrderBy(value => value.ToString());
            List<object> orderList = new List<object>();
            //foreach (object groupValue in sortedGroupValues)
            foreach (object groupValue in allowedValues)
            {
                orderList.Add(groupValue);
            }
            field.Tag = orderList;
            field.SortMode = PivotSortMode.Custom;
        }

        void RemoveSingleFieldOrder(PivotGridField field)
        {
            field.SortMode = PivotSortMode.Default;
        }

        void InitColumnsOrder()
        {
            foreach (PivotGridField field in targetPivot.GetFieldsByArea(PivotArea.ColumnArea))
            {
                InitSingleFieldOrder(field);
            }
        }
        public void Detach(PivotGridControl pivot)
        {
            foreach (PivotGridField field in targetPivot.Fields)
            {
                {
                    field.Tag = null;
                    field.SortMode = PivotSortMode.Default;
                }
            }

        }
        private void targetPivot_CustomFieldSort(object sender, DevExpress.XtraPivotGrid.PivotGridCustomFieldSortEventArgs e)
        {
            if (AllowDragRow && e.Field.Area == PivotArea.RowArea)
            {
                List<object> orderlist = e.Field.Tag as List<object>;
                //Comparer comp = new Comparer;
                if (orderlist == null || orderlist.IndexOf(e.Value1) == -1 || orderlist.IndexOf(e.Value2) == -1)
                {
                    //e.Result = Comparer.DefaultInvariant.Compare(e.Value1, e.Value2);
                    //e.Handled = true;
                    return;
                }
                e.Result = Comparer.DefaultInvariant.Compare(orderlist.IndexOf(e.Value1), orderlist.IndexOf(e.Value2));
                e.Handled = true;
            }
            if (AllowDragColumn && e.Field.Area == PivotArea.ColumnArea)
            {
                List<object> orderlist = e.Field.Tag as List<object>;
                if (orderlist == null || orderlist.IndexOf(e.Value1) == -1 || orderlist.IndexOf(e.Value2) == -1)
                {
                    return;
                }
                e.Result = Comparer.DefaultInvariant.Compare(orderlist.IndexOf(e.Value1), orderlist.IndexOf(e.Value2));
                e.Handled = true;
            }
        }
        PivotGridHitInfo dragInfo;
        PivotGridHitInfo _targetInfo;
        PivotGridHitInfo TargetInfo
        {
            get { return _targetInfo; }
            set
            {
                _targetInfo = value;
            }
        }
        bool onDrag = false;
        private void targetPivot_MouseMove(object sender, MouseEventArgs e)
        {
            if (!onDrag)
            {
                if (e.Button == System.Windows.Forms.MouseButtons.Left)
                {
                    PivotGridHitInfo hi = targetPivot.CalcHitInfo(e.Location);
                    if (hi.HitTest == PivotGridHitTest.Value)
                    {
                        if ((AllowDragRow && hi.ValueInfo.Field.Area == PivotArea.RowArea) || (AllowDragColumn && hi.ValueInfo.Field.Area == PivotArea.ColumnArea))
                        {
                            dragInfo = hi;
                            onDrag = true;
                            targetPivot.FindForm().Cursor = Cursors.Hand;
                        }
                    }
                }
            }
            else
            {
                DragDropEffects allowedEffect = GetDragEffect(e.Location);
                if (allowedEffect == DragDropEffects.None)
                    targetPivot.FindForm().Cursor = Cursors.No;
                else
                    targetPivot.FindForm().Cursor = Cursors.Hand;


            }

        }

        private void targetPivot_MouseUp(object sender, MouseEventArgs e)
        {
            if (onDrag)
            {
                if (GetDragEffect(e.Location) == DragDropEffects.Move)
                {
                    PivotGridHitInfo hi = targetPivot.CalcHitInfo(e.Location);
                    if (hi.ValueInfo.Field == dragInfo.ValueInfo.Field)
                    {
                        (dragInfo.ValueInfo.Field.Tag as IList).Remove(dragInfo.ValueInfo.Value);
                        (dragInfo.ValueInfo.Field.Tag as IList).Insert((dragInfo.ValueInfo.Field.Tag as IList).IndexOf(hi.ValueInfo.Value), dragInfo.ValueInfo.Value);
                        //updateMemos();
                    }
                    if ((dragInfo.ValueInfo.Field.Area == PivotArea.RowArea && AllowChangeRowGroup) || (dragInfo.ValueInfo.Field.Area == PivotArea.ColumnArea && AllowChangeColumnGroup))
                    {
                        PivotDrillDownDataSource ds = null;
                        for (int i = 0; i < dragInfo.ValueInfo.Field.AreaIndex && i <= hi.ValueInfo.Field.AreaIndex; i++)
                        {
                            object targetFieldValue = hi.ValueInfo.GetFieldValue(targetPivot.GetFieldByArea(dragInfo.ValueInfo.Field.Area, i), hi.ValueInfo.MinIndex);
                            object dragFieldValue = dragInfo.ValueInfo.GetFieldValue(targetPivot.GetFieldByArea(dragInfo.ValueInfo.Field.Area, i), dragInfo.ValueInfo.MinIndex);
                            if (targetFieldValue != dragFieldValue)
                            {
                                if (ds == null)
                                    ds = dragInfo.ValueInfo.CreateDrillDownDataSource();
                                foreach (PivotDrillDownDataRow row in ds)
                                {
                                    ds.SetValue(row.Index, targetPivot.GetFieldByArea(dragInfo.ValueInfo.Field.Area, i), targetFieldValue);
                                }
                            }
                        }
                    }
                    targetPivot.RefreshData();
                }

                onDrag = false;
                targetPivot.FindForm().Cursor = Cursors.Default;
            }
        }

        DragDropEffects GetDragEffect(Point point)
        {
            PivotGridHitInfo hi = targetPivot.CalcHitInfo(point);
            PivotArea dragArea = dragInfo.ValueInfo.Field.Area;
            if (hi.HitTest != PivotGridHitTest.Value || hi.ValueInfo.Field == null || hi.ValueInfo.Field.Area != dragArea)
            {
                TargetInfo = null;
                return DragDropEffects.None;
            }
            if ((dragArea == PivotArea.ColumnArea && !AllowDragColumn) || (dragArea == PivotArea.RowArea && !AllowDragRow))
            {
                TargetInfo = null;
                return DragDropEffects.None;
            }
            //for the same group
            if (IsSameGroup(dragInfo.ValueInfo, hi.ValueInfo))
            {
                if (dragInfo.ValueInfo.Field != hi.ValueInfo.Field || dragInfo.ValueInfo.Value.Equals(hi.ValueInfo.Value))
                {
                    TargetInfo = null;
                    return DragDropEffects.None;
                }
                else
                {
                    TargetInfo = hi;
                    return DragDropEffects.Move;
                }
            }
            else //for different groups
            {
                if ((dragArea == PivotArea.RowArea && AllowChangeRowGroup) || (dragArea == PivotArea.ColumnArea && AllowChangeColumnGroup))
                {
                    TargetInfo = hi;
                    return DragDropEffects.Move;
                }
                else
                {
                    TargetInfo = null;
                    return DragDropEffects.None;
                }
            }
        }

        bool IsSameGroup(PivotFieldValueHitInfo dragInfo, PivotFieldValueHitInfo targetInfo)
        {
            if (dragInfo.Field.Area != targetInfo.Field.Area)
                return false;
            int minAreaIndex = Math.Min(dragInfo.Field.AreaIndex-1, targetInfo.Field.AreaIndex);
            for (int i = 0; i <= minAreaIndex; i++)
            {
                object field1GroupValue = dragInfo.GetFieldValue(targetPivot.GetFieldByArea(dragInfo.Field.Area, i), dragInfo.MinIndex);
                object field2GroupValue = targetInfo.GetFieldValue(targetPivot.GetFieldByArea(targetInfo.Field.Area, i), targetInfo.MinIndex);
                if (!field1GroupValue.Equals(field2GroupValue))
                {
                    return false;
                }
            }
            return true;
        }


    }
}
